odl/operator/discr_ops.py

```diff
@@ -0,0 +1,446 @@
+# Copyright 2014, 2015 The ODL development group
+#
+# This file is part of ODL.
+#
+# ODL is free software: you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation, either version 3 of the License, or
+# (at your option) any later version.
+#
+# ODL is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with ODL.  If not, see <http://www.gnu.org/licenses/>.
+
+"""Operators defined on `DiscreteLp`."""
+
+# Imports for common Python 2/3 codebase
+from __future__ import print_function, division, absolute_import
+from future import standard_library
+standard_library.install_aliases()
+from builtins import super
+
+import numpy as np
+
+# ODL imports
+from odl.operator.operator import Operator
+from odl.set.pspace import ProductSpace
+from odl.discr.lp_discr import DiscreteLp
+
+
+__all__ = ('DiscretePartDeriv', 'DiscreteGradient', 'DiscreteDivergence')
+
+
+def finite_diff(f, out=None, axis=0, dx=1.0, edge_order=None,
+                zero_padding=False):
+    """Calculate the partial derivative of `f` along a given `axis` using
+    discrete differences.
+
+    The partial derivative is computed using second-order accurate central
+    differences in the interior and either first- or second-order accurate
+    one-sides (forward or backward) differences at the boundaries.
+
+    Assuming (implicit) zero padding central differences are used on the
+    interior and on endpoints. Otherwise one-sided differences are used. If
+    `zero_padding` is False first-order accuracy can be triggered on
+    endpoints with parameter `edge_order`.
+
+    The returned array has the same shape as the input array `f`.
+
+    Parameters
+    ----------
+    f : array-like
+         An N-dimensional array
+    out : array-like
+         An N-dimensional array
+    axis : `int`, optional
+        The axis along which the partial derivative is evaluated. Default: 0
+    dx : `float`, optional
+        Scalars specifying the sample distances in each dimension `axis`.
+        Default distance: 1.0
+    edge_order : {1, 2}, optional
+        First-order accurate differences (1) can be used at the boundaries
+        if no zero padding is used. Default edge order: 2
+    zero_padding : `bool`, optional
+        Implicit zero padding. Assumes values outside the domain of `f` to be
+        zero. Default: False
+
+    Returns
+    -------
+    out : `numpy.ndarray`
+        Returns an N-dimensional array of the same shape as `f`
```

Comment:
Simply `make html` in the doc folder. You may need to run the `generate_doc.py` script in doc/source first to populate the files etc.